### PR TITLE
ENH: Implement Morton distance

### PR DESCRIPTION
**Pull request by tastatham**
Implement morton_distance for Dask-GeoPandas. This provides another way of spatially partitioning Dask-DataFrames using another space-filling curve. 

Details:
* Calculating the hilbert distance will allow us to partition Dask-GeoPandas using this information.
* This supports both GeoPandas & Dask-GeoPandas
* The output is returned as a Pandas Series

When compared to the Hilbert distance, there is very little difference in compute time. However, from some basic comparisons, the Hilbert distance does appear to preserve locality better. Further work needs to be carried out to compare the two (as well as a third partitioning method; Geohash).

### CODE REVIEW

File: continuous_integration/envs/37-latest.yaml [hunk 1]
```diff
@@ -14,3 +14,4 @@ dependencies:
   - pytest
   - pytest-cov
   - hilbertcurve
+  - pymorton
```
Comment (martinfleis) [inline continuous_integration/envs/37-latest.yaml:17]:
```suggestion
  - pip
  - pip:
    - pymorton
```

`pymorton` seems to be available only on PyPI. You'll need to do the same change in all envs.

File: dask_geopandas/morton_distance.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
+import numpy as np
+import pandas as pd
+from dask_geopandas.hilbert_distance import _continuous_to_discrete_coords
+
+# Based on #https://code.activestate.com/recipes/
+# 577558-interleave-bits-aka-morton-ize-aka-z-order-curve/
```
Comment (martinfleis) [inline dask_geopandas/morton_distance.py:6]: The original code is licensed under LGPL 2.1. Can you ensure we can use it here and under which conditions?

File: dask_geopandas/morton_distance.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+    gdf : GeoDataFrame
+
+    total_bounds : Total bounds
+
+    p : number of iterations
+
+    """
+    bounds = gdf.bounds.to_numpy()
+    coords = _continuous_to_discrete_coords(total_bounds, bounds, p)
```
Comment (martinfleis) [inline dask_geopandas/morton_distance.py:24]: The `p` parameter is the same as the Hilbert curve has? I mean, denoting the same concept?

**Review by martinfleis — COMMENTED** [review]
Thanks a lot! First quick pass (haven't played with it yet).

Can it be combined with numba? That loop in `_distances_from_coordinates` may benefit from that with large GeoDataFrames.

File: dask_geopandas/morton_distance.py (hunk 2 above)
Comment (jorisvandenbossche) [reply dask_geopandas/morton_distance.py:6]: If it's LGPL, we can't use it here. However, the recipe states that it's itself based on other code, which I think we can probably use if we refer directly to those.

File: dask_geopandas/morton_distance.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+def part1by1(n):
+
+    """Interleave bits by Binary Magic Numbers"""
+
+    n &= 0x0000FFFF
+    n = (n | (n << 8)) & 0x00FF00FF
+    n = (n | (n << 4)) & 0x0F0F0F0F
+    n = (n | (n << 2)) & 0x33333333
+    n = (n | (n << 1)) & 0x55555555
```
Comment (jorisvandenbossche) [inline dask_geopandas/morton_distance.py:74]: Currently you work on scalar numbers? I think those bitwise operations should normally also work vectorized on the full array?

File: dask_geopandas/morton_distance.py [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    Parameters
+    ----------
+
+    coords : Coordinates in np.array format
+    """
+
+    result = np.zeros(coords.shape[0], dtype=np.int64)
+    # For each coord calculate hilbert distance
+    for i in range(coords.shape[0]):
+        result[i] = encode_morton(coords[i][0], coords[i][1])
```
Comment (jorisvandenbossche) [inline dask_geopandas/morton_distance.py:45]:
```suggestion
    result = encode_morton(coords[:, 0], coords[:, 1])
```

(see comment below about this (probably) working on full numpy arrays as well)

**Review by jorisvandenbossche — COMMENTED** [review]
Naming question: if I google "morton distance", I basically get no useful result (while "hilbert distance" does give something useful). For google it, you rather need something like "morton curve". So wondering if "morton distance" is the best name (although it's of course the "distance along the curve", I suppose? 
Now, that still results in a wiki page for which the main title is "Z-order curve", so we could also consider using "Z-order" instead of "Morton" for the naming?

**Comment by jorisvandenbossche** [comment]
> When compared to the Hilbert distance, there is very little difference in compute time. 

I suppose that with using the numpy-based implementation, this will become faster. A
lso, best test the performance on somewhat larger data (eg I see in your notebook that you are timing it on the "naturalearth_lowres" data, which is tiny. If you profile that timing (eg with snakeviz), you'll see that you are mostly measuring overhead, and the actual hilbert/morton encoding only is a small part). You can also easily combine the data multiple times to artificially get bigger data (`gdf=pd.concat([gdf]*1000, axis=0)`)

File: dask_geopandas/morton_distance.py (hunk 2 above)
Comment (jorisvandenbossche) [reply dask_geopandas/morton_distance.py:6]: Also, https://github.com/trevorprater/pymorton has similar code and which is MIT, so that's fine. So in general this code (the specific bit values) seems widespread, so maybe no explicit reference is needed.

File: continuous_integration/envs/37-latest.yaml (hunk 1 above)
Comment (tastatham) [reply continuous_integration/envs/37-latest.yaml]: updated & failed in 937bbbf (39-latest.yml) ... & 2c90399. I tried adding pymorton explicitly but not sure why it's failing?

File: continuous_integration/envs/39-dev.yaml [hunk 6]
```diff
@@ -15,10 +15,11 @@ dependencies:
   - pytest
   - pytest-cov
   - hilbertcurve
+  - pymorton
```
Comment (martinfleis) [inline continuous_integration/envs/39-dev.yaml:18]: Remove this one.

File: continuous_integration/envs/39-dev.yaml [hunk 7]
```diff
@@ -15,10 +15,11 @@ dependencies:
...
   # optional dependencies
   - pyarrow
-  - pip
   - pip:
+    - git+https://github.com/trevorprater/pymorton.git@master
```
Comment (martinfleis) [inline continuous_integration/envs/39-dev.yaml:22]: Keep `pip` in there and install `pymorton` from PyPI like in the other envs. We don't need its master here.

File: continuous_integration/envs/39-dev.yaml (hunk 6 above)
Comment (tastatham) [reply continuous_integration/envs/39-dev.yaml]: > Remove this one.

ah, I just forgot to remove it as a conda dependency :facepalm:

updated in 72bb05d

File: dask_geopandas/morton_distance.py (hunk 3 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py:24]: I believe the following logic holds true - correct me otherwise;
Using the `_continuous_to_discrete_coords` transforms each of the coordinate pairs from continuous to integers.
The p (or order of curve) value will control the number of iterations in constructing the [Hilbert/Morton] curve. 
Then calculated distances along each curve work entirely in terms of integers, where the size of the hypercube will reflect the number of bits per dimension - controlled by p;

* The 1st order curve supports values (0|1) on both x and y axis - giving us 2 bit binary numbers of the range from 0,1,2,3.
* Then the 2nd order curve supports values (0|1|2|3) on ... - giving us 4 bit binary numbers of the range 0,1...,15 and so on...

Therefore, using the same p value (or n iterations) for calculated distances along both curves will be bound by the same size hypercube.
 
The maximum discrete distances along each order are as follows;
```python
iterations = 16 # or p 
n = 2 # two-dimension
max_discrete_distances = []
for p in range(iterations):
    max_discrete_distances.append(2 ** (n*p)  - 1 ) # -1 to start index at 0
max_discrete_distances[1:]  # from 1st to 15th order curve
```

long story short, yes it denotes the same concept.

Comment (martinfleis) [reply dask_geopandas/morton_distance.py:24]: Cool! Thanks for the great explanation!

**Comment by tastatham** [comment]
> Naming question: if I google "morton distance", I basically get no useful result (while "hilbert distance" does give something useful). For google it, you rather need something like "morton curve". So wondering if "morton distance" is the best name (although it's of course the "distance along the curve", I suppose?
> Now, that still results in a wiki page for which the main title is "Z-order curve", so we could also consider using "Z-order" instead of "Morton" for the naming?

I did think about this briefly and thought that the naming of the `morton_distance` was appropriate. This is because;
*  We are calculating the distance along the Morton/Z-order curve. 
* To not confuse Z-order with Z-dimension
*  It's consistent with the `hilbert_distance` function (which we can also change)

I'm happy to discuss this later on in the call.

File: dask_geopandas/morton_distance.py (hunk 4 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py:74]: > Currently you work on scalar numbers? I think those bitwise operations should normally also work vectorized on the full array?
Yes, I can update this later to use numpy.

**Comment by tastatham** [comment]
> > When compared to the Hilbert distance, there is very little difference in compute time.
> 
> I suppose that with using the numpy-based implementation, this will become faster. A
> lso, best test the performance on somewhat larger data (eg I see in your notebook that you are timing it on the "naturalearth_lowres" data, which is tiny. If you profile that timing (eg with snakeviz), you'll see that you are mostly measuring overhead, and the actual hilbert/morton encoding only is a small part). You can also easily combine the data multiple times to artificially get bigger data (`gdf=pd.concat([gdf]*1000, axis=0)`)

Good point, I will profile this and the numpy version using the artificial dataset.

File: dask_geopandas/morton_distance.py (hunk 3 above)
Comment (jorisvandenbossche) [reply dask_geopandas/morton_distance.py:24]: > The p (or order of curve) value will control the number of iterations in constructing the [Hilbert/Morton] curve.

But as far as I can see, to calculate the Morton curve, we are not using any iterations? (in contrast to Hilbert curve) 
The `encode_morton` and `part1by1` seem to do a fixed set of computations independent of `p`.

(the explanation that it controls the size of the hypercube (and thus the precision of the calculated distances) is of course still valid nonetheless)

Comment (jorisvandenbossche) [reply dask_geopandas/morton_distance.py:24]: (I suppose the "classical" implementation also uses loops, but the bit-shifting method is an optimized version that doesn't require loops)

File: dask_geopandas/core.py [hunk 8]
```diff
@@ -336,6 +337,33 @@ def hilbert_distance(self, p=15):
 
         return distances
 
+    def morton_distance(self, p=15):
+
+        """
+
```
Comment (jorisvandenbossche) [inline dask_geopandas/core.py:343]: TODO: expand the docstring (quick explanation, mention "also known as", maybe link to wiki page)

File: dask_geopandas/core.py [hunk 9]
```diff
@@ -336,6 +337,33 @@ def hilbert_distance(self, p=15):
...
+        Parameters
+        ----------
+
+        p : The number of iterations used in constructing the Hilbert curve.
```
[the remark this answers is not in the document]
Comment (brendan-ward) [reply dask_geopandas/core.py:347]: Here and below, see also the [numpydoc](https://numpydoc.readthedocs.io/en/latest/format.html#parameters) standard for parameters in the docstring
```
param : type
    description
```

File: dask_geopandas/morton_distance.py (hunk 4 above)
Comment (brendan-ward) [reply dask_geopandas/morton_distance.py:73]: Provided I translated to numpy correctly:
```
n = np.bitwise_and(n, 0x0000FFFF)
n = np.bitwise_and(np.bitwise_or(n, np.left_shift(n, 8)), 0x00FF00FF)
n = np.bitwise_and(np.bitwise_or(n, np.left_shift(n, 4)), 0x0F0F0F0F)
n = np.bitwise_and(np.bitwise_or(n, np.left_shift(n, 2)), 0x33333333)
n = np.bitwise_and(np.bitwise_or(n, np.left_shift(n, 1)), 0x55555555)
```

File: dask_geopandas/core.py [hunk 10]
```diff
@@ -336,6 +337,33 @@ def hilbert_distance(self, p=15):
...
+
+        Returns
+        ----------
+        Distances for each partition
+        """
+
+        # Compute total bounds of all partitions rather than each partition
+        total_bounds = self.total_bounds
+
+        # Calculate hilbert distances for each partition
```
Comment (brendan-ward) [inline dask_geopandas/core.py:357]:
```suggestion
        # Calculate Morton distances for each partition
```

**Review by brendan-ward — CHANGES_REQUESTED** [review]
Good progress!  The main changes are to update the docstrings.

If possible, it would be good to add the numpy version of the bitwise operations; hopefully that speeds it up significantly compared to Hilbert.

File: dask_geopandas/morton_distance.py (hunk 4 above)
Comment (martinfleis) [reply dask_geopandas/morton_distance.py:73]: @brendan-ward do we actually need this? On line 44 we pass numpy arrays and I'd guess that lines 69 to 73 are translated to numpy vectorised ops in numpy.

Comment (brendan-ward) [reply dask_geopandas/morton_distance.py:73]: Nope, you are right, they already use the numpy bitwise operators.

File: dask_geopandas/morton_distance.py [hunk 11]
```diff
@@ -0,0 +1,75 @@
+import numpy as np
+import pandas as pd
+from dask_geopandas.hilbert_distance import _continuous_to_discrete_coords
```
[the remark this answers is not in the document]
Comment (tastatham) [reply dask_geopandas/morton_distance.py:3]: yes I agree, both `_continuous_to_discrete_coords` and it's dependency `_continuous_to_discrete` could be centralized (outside of hilbert_distance.py) 

My initial thought was that this didn't warrant a standalone script. However, both `_hilbert_distance` and `_morton_distance` also both use `_distances_from_coordinates` in [[1]](https://github.com/tastatham/dask-geopandas/blob/a1eae6ceb8cd580f3d1f8240d22e52a850c1e5bf/dask_geopandas/hilbert_distance.py#L112)
 and [[2]](https://github.com/tastatham/dask-geopandas/blob/a1eae6ceb8cd580f3d1f8240d22e52a850c1e5bf/dask_geopandas/morton_distance.py#L31) - although this is used in slightly different ways. 

Therefore it would make sense to centralize both `_continuous_to_discrete_coords` and (a generalised version of) `_distances_from_coordinates` (to work with both `_hilbert_distance` and `_morton_distance`). Does anyone have any sugestions where to centralize these functions?

File: dask_geopandas/morton_distance.py [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    Calculate Morton distances from coordinates
+
+    Parameters
+    ----------
+
+    coords : Coordinates in np.array format
+    """
+
+    result = np.zeros(coords.shape[0], dtype=np.int64)
```
[the remark this answers is not in the document]
Comment (tastatham) [reply dask_geopandas/morton_distance.py]: thanks! I realised this was still in there after the commit. Updated in 703b5b4

File: dask_geopandas/core.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (tastatham) [reply dask_geopandas/core.py:347]: @brendan-ward thanks for your suggestion about formatting docstring. I have updated this in 703b5b4

File: dask_geopandas/core.py (hunk 8 above)
Comment (tastatham) [reply dask_geopandas/core.py:343]: Updated in 703b5b4

File: continuous_integration/envs/39-dev.yaml (hunk 7 above)
Comment (tastatham) [reply continuous_integration/envs/39-dev.yaml]: resolved in 9756156

File: dask_geopandas/morton_distance.py (hunk 4 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py:79]: I tried both but I found the original function to be easier to read.

File: dask_geopandas/morton_distance.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (martinfleis) [reply dask_geopandas/morton_distance.py:3]: Can you make a note of it in the hilbetr follow-up issue (#85) and keep it as it is in this PR? We'll clean this later (I would even say that both `morton` and `hilbert` py files should go to `repartition` folder or something like that to keep the file structure clean.

File: continuous_integration/envs/39-dev.yaml [hunk 13]
```diff
@@ -17,8 +17,8 @@ dependencies:
   - hilbertcurve
   # optional dependencies
   - pyarrow
-  - pip
```
Comment (martinfleis) [inline continuous_integration/envs/39-dev.yaml:20]: We should always have `pip` listed among dependencies if we use it to install some. In all envs with `pymorton`. Conda adds it automatically but complains about it.

File: dask_geopandas/core.py (hunk 8 above)
Comment (martinfleis) [reply dask_geopandas/core.py:343]: @tastatham This is a top-level API function (only one in this PR) so this one needs a proper docstring since this will show up in the API docs (once they're done). We need a full docstring here. Everything else is internal. This even looks that you have removed part of the docstring from here.

File: dask_geopandas/core.py [hunk 14]
```diff
@@ -336,6 +337,33 @@ def hilbert_distance(self, p=15):
...
+
+        p : The number of iterations used in constructing the Hilbert curve.
+
+        Returns
+        ----------
+        Distances for each partition
+        """
+
+        # Compute total bounds of all partitions rather than each partition
+        total_bounds = self.total_bounds
```
Comment (martinfleis) [inline dask_geopandas/core.py:355]: When we update hilbert to allow custom bounds, we should do the same here. Can you make a public note of it somewhere?

File: dask_geopandas/core.py (hunk 8 above)
Comment (jorisvandenbossche) [reply dask_geopandas/core.py:343]: You can copy the docstring you added to `_morton_distance` here

File: dask_geopandas/morton_distance.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+    type : int
+        Integer distances from Morton curve
+    """
+
+    return np.bitwise_or(
+        _part1by1(coords[:, 0]), np.left_shift(_part1by1(coords[:, 1]), 1)
+    )
+
+
+def _part1by1(n):
```
Comment (jorisvandenbossche) [inline dask_geopandas/morton_distance.py:59]: Do you have an idea where the "part1by1" name comes from? (I mean, it comes from the active state recipe, but what's the reason for it? (I find it a bit cryptic name))

File: dask_geopandas/morton_distance.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+    n : np.array
+        X or Y coordinates
+
+    Returns
+    -------
+    n : int
+        Interleaved bits
+    """
+
+    n &= 0x0000FFFF
```
Comment (jorisvandenbossche) [inline dask_geopandas/morton_distance.py:74]: This line is not included in the code in http://graphics.stanford.edu/~seander/bithacks.html#InterleaveBMN (the reference you added now). Do you know what it is for? (it's eg also there in pymorton https://github.com/trevorprater/pymorton/blob/f248683c19d90e904193c58bbd03e77bc2c43768/pymorton/pymorton.py#L11)

File: dask_geopandas/morton_distance.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
+import numpy as np
+import pandas as pd
+from dask_geopandas.hilbert_distance import _continuous_to_discrete_coords
+
+# Based on #http://graphics.stanford.edu/~seander/bithacks.html#InterleaveBMN
```
Comment (jorisvandenbossche) [inline dask_geopandas/morton_distance.py:5]: I would move this comment to the function that does the bit-shifting (since it's only that part that is based on that)

File: dask_geopandas/core.py (hunk 14 above)
Comment (tastatham) [reply dask_geopandas/core.py:355]: I've opened a follow up issue in #92

File: dask_geopandas/morton_distance.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (tastatham) [reply dask_geopandas/morton_distance.py:3]: Updated issue #85 with a [new comment](https://github.com/geopandas/dask-geopandas/issues/85#issuecomment-894667314)

File: dask_geopandas/morton_distance.py (hunk 16 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py]: yes, it's also not used in the  [`Real-Time Collision Detection book, Chapter 7, 317`](http://www.r-5.org/files/books/computers/algo-list/realtime-3d/Christer_Ericson-Real-Time_Collision_Detection-EN.pdf

Updated in 27ca9e7

File: dask_geopandas/morton_distance.py (hunk 17 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py]: Updated in 27ca9e7

File: dask_geopandas/morton_distance.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        x, y coordinate pairs based on mid-points of geoms
+
+    Returns
+    -------
+    type : int
+        Integer distances from Morton curve
+    """
+
+    return np.bitwise_or(
+        _part1by1(coords[:, 0]), np.left_shift(_part1by1(coords[:, 1]), 1)
```
[the remark this answers is not in the document]
Comment (tastatham) [reply dask_geopandas/morton_distance.py]: Updated in 27ca9e7

File: dask_geopandas/morton_distance.py (hunk 15 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py:54]: I agree, the naming convention is a bit cryptic but seems to be widely adopted. See [`Real-Time Collision Detection book, Chapter 7`](http://www.r-5.org/files/books/computers/algo-list/realtime-3d/Christer_Ericson-Real-Time_Collision_Detection-EN.pdf
).

I think it comes from the fact the part1by1 function inserts one zero bit between each original data bit (part), the part1by2 function inserts two zero bits between each original data bit (part) and so on...

File: dask_geopandas/core.py (hunk 8 above)
Comment (tastatham) [reply dask_geopandas/core.py:346]: updated in 2c539e0

File: continuous_integration/envs/39-dev.yaml (hunk 13 above)
Comment (tastatham) [reply continuous_integration/envs/39-dev.yaml:20]: Updated in 07fdecc

File: dask_geopandas/morton_distance.py (hunk 16 above)
Comment (jorisvandenbossche) [reply dask_geopandas/morton_distance.py]: Looking a little bit into it, this might actually still be needed because we are using int64 for the integer coordinates. With the current value of `p=15` that you are using for testing, the coordinates never become larger than the max value of an uint16. 

But if you try to run the tests with eg `p=20`, do they still pass?

The example code in the book you reference uses `uint32` type for the input values, and says it only uses the low 16 bits of those numbers (and then spreading those out with the bit interleaving). So if I am understanding correctly, if your coordinates are 16bit, you view then as 32bit when passing to the bit interleaving code. And that line `n &= 0x0000FFFF` might ensure that it only uses the lower 16 bits? (although I still don't fully understand why it is in there, as if you have a number that is larger than what can be represented by the 16bits (>65535), it gets truncated by `n &= 0x0000FFFF` anyway and then you also get a wrong result ..
But eg GEOS also has a Morton implementation, and includes it as well: https://github.com/libgeos/geos/blob/16610691611decd2b9545ea68ef41e8b94f859e6/src/shape/fractal/MortonCode.cpp#L95-L104

File: continuous_integration/envs/38-latest.yaml [hunk 19]
```diff
@@ -14,5 +14,7 @@ dependencies:
   - pytest
   - pytest-cov
   - hilbertcurve
+  - pip:
+    - pymorton
   # optional dependencies
   - pyarrow
```
Comment (martinfleis) [inline continuous_integration/envs/38-latest.yaml:20]:
```suggestion
  # optional dependencies
  - pyarrow
  - pip
  - pip:
    - pymorton
```

It is better to keep pip dependencies on the bottom. Also, as I mentioned before, try to always list `pip` if there are pip installed dependencies.

File: continuous_integration/envs/39-latest.yaml [hunk 20]
```diff
@@ -14,5 +14,7 @@ dependencies:
   - pytest
   - pytest-cov
   - hilbertcurve
+  - pip:
+    - pymorton
   # optional dependencies
   - pyarrow
```
Comment (martinfleis) [inline continuous_integration/envs/39-latest.yaml:20]:
```suggestion
  # optional dependencies
  - pyarrow
  - pip
  - pip:
    - pymorton
```

File: continuous_integration/envs/37-latest.yaml [hunk 21]
```diff
@@ -14,3 +14,5 @@ dependencies:
   - pytest
   - pytest-cov
   - hilbertcurve
+  - pip:
```
Comment (martinfleis) [inline continuous_integration/envs/37-latest.yaml:17]:
```suggestion
  - pip
  - pip:
```

**Review by martinfleis — COMMENTED** [review]
Final super tiny suggestion from me. Otherwise, this is good to go. Nice job @tastatham!

File: dask_geopandas/morton_distance.py (hunk 16 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py]: yes, that's correct. These examples assume the input values are of type `uint32`. Currently `uint64` coords are passed to the `_part1by1` function. This comes from the `_continuous_to_discrete_coords` function. 

In #70 and #90 we discussed several follow up issues; 
1. Setting the dtype for the `_continuous_to_discrete_coords` function
2. Centralizing several functions which are used by both Hilbert and Morton distance.  

We could look into this further alongside these follow up issues and for now, simply pass the coords to the `_part1by1` function as type `unit32`? - coords.astype("uint32")

File: dask_geopandas/core.py [hunk 22]
```diff
@@ -336,6 +337,38 @@ def hilbert_distance(self, p=15):
...
+        Parameters
+        ----------
+
+        p : int
+            precision of the Morton curve
+
+        Returns
+        ----------
+        type : pandas.Series
+            Series containing distances from Morton curve
```
Comment (martinfleis) [inline dask_geopandas/core.py:356]: It returns dask Series, not pandas Series.

Comment (tastatham) [reply dask_geopandas/core.py]: updated in a08e85e

File: dask_geopandas/core.py [hunk 23]
```diff
@@ -343,6 +344,38 @@ def hilbert_distance(self, p=15):
 
         return distances
 
+    def morton_distance(self, p=15):
+
+        """
+        Calculate distance of geoms along Morton curve
```
Comment (martinfleis) [inline dask_geopandas/core.py:350]:
```suggestion
        Calculate the distance of geometries along the Morton curve
        
        The Morton distance can be used to spatially partition Dask-GeoPandas objects,
        by mapping two-dimensional geometries along the Morton space-filing curve.
        
        Each geometry is represented by the midpoint of its bounds and linked to the 
        Morton curve. The function returns a distance from the beginning 
        of the curve to the linked point.
        
        Morton distance is more performant than ``hilbert_distance`` but can result in 
        less optimal partitioning.
```

It is better to have all info in the main docstring.

File: dask_geopandas/core.py [hunk 24]
```diff
@@ -343,6 +344,38 @@ def hilbert_distance(self, p=15):
...
+        Parameters
+        ----------
+
+        p : int
+            precision of the Morton curve
+
+        Returns
+        ----------
+        type : dask.Series
+            Series containing distances from Morton curve
```
Comment (martinfleis) [inline dask_geopandas/core.py:363]:
```suggestion
            Series containing distances along the Morton curve
```

File: dask_geopandas/core.py (hunk 23 above)
Comment (tastatham) [reply dask_geopandas/core.py]: thank you

File: dask_geopandas/morton_distance.py (hunk 16 above)
Comment (tastatham) [reply dask_geopandas/morton_distance.py]: I think this is the last component that needs updating prior to merging it. Any thoughts?

Comment (jorisvandenbossche) [reply dask_geopandas/morton_distance.py]: I would maybe revert the removal of this line for now, and then list it as a follow-up to check / test more.

Otherwise, I would at least eg add a test with a higher p value as mentioned above.

**Comment by jorisvandenbossche** [comment]
The linting is failing (probably something white space in the suggested edit)

File: dask_geopandas/core.py [hunk 25]
```diff
@@ -343,6 +344,48 @@ def hilbert_distance(self, p=15):
...
+        by mapping two-dimensional geometries along the Morton space-filing curve.
+        
+        Each geometry is represented by the midpoint of its bounds and linked to the 
+        Morton curve. The function returns a distance from the beginning 
+        of the curve to the linked point.
+        
+        Morton distance is more performant than ``hilbert_distance`` but can result in 
+        less optimal partitioning.
+
+        The Morton curve is also known as Z-order https://en.wikipedia.org/wiki/Z-order
```
Comment (jorisvandenbossche) [inline dask_geopandas/core.py:362]:
```suggestion
        The Morton curve is also known as Z-order https://en.wikipedia.org/wiki/Z-order.

        The Morton distance can be used to spatially partition Dask-GeoPandas objects,
        by mapping two-dimensional geometries along the Morton space-filing curve.
        
        Each geometry is represented by the midpoint of its bounds and linked to the 
        Morton curve. The function returns a distance from the beginning 
        of the curve to the linked point.
        
        Morton distance is more performant than ``hilbert_distance`` but can result in 
        less optimal partitioning.
```

(I think it's good to "drop" the alternative name in the beginning for visibility)

**Review by martinfleis — APPROVED** [review]
I think we're done here. Thanks a lot @tastatham!

**Review by brendan-ward — APPROVED** [review]
Thanks for working on this!

Eventually it would be good to add more guidance or at least an appropriate range for the `p` parameter, at least so that the user doesn't overflow data types by using too large a `p` parameter.

File: dask_geopandas/morton_distance.py [hunk 26]
```diff
@@ -0,0 +1,76 @@
+import pandas as pd
+from dask_geopandas.hilbert_distance import _continuous_to_discrete_coords
+
+
+def _morton_distance(gdf, total_bounds, p):
+    """
+    Calculate distance of geoms along Morton curve
```
[the remark this answers is not in the document]
Comment (tastatham) [reply dask_geopandas/morton_distance.py:7]: thanks Brendan, yes I agree. Hopefully I can find some time to do this before Monday but if not, I will definitely follow up on this.

**Comment by jorisvandenbossche** [comment]
I added back one line in the algorithm, so it matches what pymorton does (see the discussion above https://github.com/geopandas/dask-geopandas/pull/90#discussion_r684612717). We can then properly figure out if it is needed / if we need to cast the integer coordinates to a certain bitsize etc in a follow-up.

**Comment by jorisvandenbossche** [comment]
Thanks @tastatham !
